recipe/link-petsc.patch

```diff
@@ -0,0 +1,76 @@
+From a3fc5bc344905c849c8ffcdf4153bacab376c71b Mon Sep 17 00:00:00 2001
+From: Min RK <benjaminrk@gmail.com>
+Date: Mon, 22 Jan 2024 12:15:07 +0100
+Subject: [PATCH] only link direct dependencies
+
+petsc, blas, lapack, scalapack
+
+and don't export more than -lslepc in SLEPC_LIB
+---
+ config/package.py        | 3 ++-
+ config/packages/slepc.py | 5 +++--
+ gmakefile                | 6 +++---
+ 3 files changed, 8 insertions(+), 6 deletions(-)
+
+diff --git a/config/package.py b/config/package.py
+index 86c82d84b..751c1de94 100644
+--- a/config/package.py
++++ b/config/package.py
+@@ -366,8 +366,9 @@ Downloaded package %s from: %s is not a tarball.
+       self.log.Exit('Cannot create temporary directory')
+     try:
+       with open(os.path.join(tmpdir,'makefile'),'w') as makefile:
++        makefile.write("SLEPC_PETSC_SNES_LIB = -lpetsc -lblas -llapack -lscalapack\n")
+         makefile.write('checklink: checklink.o\n')
+-        makefile.write('\t${CLINKER} -o checklink checklink.o ${LINKFLAGS} ${PETSC_SNES_LIB}\n')
++        makefile.write('\t${CLINKER} -o checklink checklink.o ${LINKFLAGS} ${SLEPC_PETSC_SNES_LIB}\n')
+         makefile.write('\t@${RM} -f checklink checklink.o\n')
+         makefile.write('include '+os.path.join('${PETSC_DIR}','lib','petsc','conf','variables')+'\n')
+         makefile.write('include '+os.path.join('${PETSC_DIR}','lib','petsc','conf','rules')+'\n')
+diff --git a/config/packages/slepc.py b/config/packages/slepc.py
+index f6f8e79d6..86c2f61b1 100644
+--- a/config/packages/slepc.py
++++ b/config/packages/slepc.py
+@@ -62,10 +62,11 @@ class SLEPc(package.Package):
+     # Single library installation
+     if petsc.singlelib:
+       slepcvars.write('SHLIBS = libslepc\n')
++      slepcvars.write('SLEPC_PETSC_SNES_LIB = -L${SLEPC_LIB_DIR} -lpetsc -lblas -llapack -lscalapack\n')
+       slepcvars.write('LIBNAME = '+os.path.join('${INSTALL_LIB_DIR}','libslepc.${AR_LIB_SUFFIX}')+'\n')
+       for module in ['SYS','EPS','SVD','PEP','NEP','MFN','LME']:
+-        slepcvars.write('SLEPC_'+module+'_LIB = ${CC_LINKER_SLFLAG}${SLEPC_LIB_DIR} -L${SLEPC_LIB_DIR} -lslepc ${SLEPC_EXTERNAL_LIB} ${PETSC_SNES_LIB}\n')
+-      slepcvars.write('SLEPC_LIB = ${CC_LINKER_SLFLAG}${SLEPC_LIB_DIR} -L${SLEPC_LIB_DIR} -lslepc ${SLEPC_EXTERNAL_LIB} ${PETSC_SNES_LIB}\n')
++        slepcvars.write('SLEPC_'+module+'_LIB = ${CC_LINKER_SLFLAG}${SLEPC_LIB_DIR} -L${SLEPC_LIB_DIR} -lslepc -lpetsc\n')
++      slepcvars.write('SLEPC_LIB = ${CC_LINKER_SLFLAG}${SLEPC_LIB_DIR} -L${SLEPC_LIB_DIR} -lslepc -lpetsc\n')
```

Comment:
note: run_exports should match whatever is here. This matches `pkg-config --libs slepc`, which produces:

```
-L/Users/minrk/conda/envs/debug-dolfinx/lib -lslepc -lpetsc
```

and where pkg-config is used (e.g. dolfinx), the results seem to be a lot more reliable than when `petscvariables` and friends are used, which dramatically over-link everything.

Reply:
The only thing I do not like is your hardwiring of `-lslepc -lpetsc` etc. Maybe you can use the proper makefiles variables for that:

`-lslepc` -> `${SLEPC_LIB_BASIC}` 

`-lpetsc` -> `${PETSC_SNES_LIB_BASIC}` 

`-lblas -llapack` -> `${BLASLAPACK_LIB}`

`-lscalapack` -> `${SCALAPACK_LIB}`

Those variables come from PETSc/SLEPc makefile support, therefore they should be defined.

Reply:
Sure, thanks for the tips! FWIW, PETSC_SNES_LIB_BASIC is not defined. PETSC [only defines the basic vars for KSP, TS, and TAO](https://gitlab.com/petsc/petsc/-/blob/v3.20.3/config/PETSc/Configure.py#L458-461) for some reason.

I would have done that if I had any capability of writing a patch that could be submitted upstream, but I don't think I will ever have sufficient knowledge of petsc's build system to do that. I'm way more comfortable maintaining a patch that 'works for us' here than a patch that won't break anything upstream.

Reply:
updated this to use the variables specified, with the exception of using `${PETSC_LIB_BASIC}` where `${PETSC_SNES_LIB_BASIC}` should be used, but isn't defined.

Reply:
I think that `${PETSC_SNES_LIB_BASIC}` is always [defined in the makefiles](https://gitlab.com/petsc/petsc/-/blob/v3.20.3/lib/petsc/conf/variables#L108), the code in `Configure.py` would just override it in the case of 'single library' but it is not necessary.

Reply:
True. It is defined, but needs an _override_ for the single-library case [like the others](https://gitlab.com/petsc/petsc/-/blob/v3.20.3/config/PETSc/Configure.py#L458-461), because has the wrong value if it's not overridden: `-lpetscsnes -lpetsc`, which results in `petscsnes not found` if anyone tries to use it.